include/tvm/runtime/ndarray.h

```diff
@@ -156,23 +156,26 @@ class NDArray : public ObjectRef {
   TVM_DLL static NDArray Empty(ShapeTuple shape, DLDataType dtype, Device dev,
                                Optional<String> mem_scope = NullOpt);
   /*!
-   * \brief Create a NDArray backed by an external DLTensor.
+   * \brief Try to create a NDArray backed by an external DLTensor without copying.
    *
-   * This allows us to create a NDArray using the memory
-   * allocated by an external source. Responsibility for memory
-   * retaining lies with the external source.
-   * \param dl_tensor The DLTensor to copy from.
+   * Fail if DLTensor is not contiguous.
+   * If AbilityOfZeroCopyForDLTensor is true a NDArray is created
+   * using the memory allocated by an external source.
+   * Responsibility for memory retaining lies with the external source.
+   * Otherwise new NDArray is created, the data is copied from the DLTensor.
```

Comment:
This says data will be copied but the `\brief` says data will not be copied. Which is it? My reading of the code below is that data will be copied if `AbilityOfZeroCopyForDLTensor` is `false`.

I don't think we should change the semantics of this function. How about we just raise an error if the `AbilityOfZeroCopyForDLTensor` is `false`?

Reply:
No, we do not raise the error due to the following reason. If you see in detail it is development of 'set_input' method of VirtualMachine. This method can be separated on two parts: one considers input as NDArray, another one does input as DLTensor. In both cases it trys to do zero copy if can, otherwise real copy is used. It has been implemented for NDArray (it automatically does not have problem with alignment) and after we have developed it for DLTensor but not checked alignment in previous PR. If we cannot use zero copy we still should use usual copy to avoid method failure. There are no 'set_input' and 'set_input_zero_copy' methods for VM. I discussed it on previous PR, it is design of VM. It means that 'set_input' should work stably in both cases with and without copying

Reply:
What you are proposing here is a change to the semantics of `FromExternalDLTensor`. Because this is a public and important API I suggest you do not change it and instead put the logic into `set_input`.

If NDArray needs its underlying DLTensor to be aligned, then we should also add a check to FromExternalDLTensor to make sure that the underlying data is aligned.

@tqchen maybe you can provide some feedback here.

Reply:
I agree that FromExternalDLTensor in this case should check alignment,